### DISCUSSION TITLE
Should we also have full generated playlists without filtering alternatives?

### DISCUSSION THREAD

**Discussion by AppleSheeple** [iptv-org/community#995] [Ideas]
Generated playlists seem to operate on the assumption that a channel must have one canonical stream that will always be the best (stability, availability, quality), while alternatives are of some lower importance.

This is all the *"fault"* of this line:

``` ts
streams = streams.uniqBy((stream: Stream) => (stream.channel || _.uniqueId()) + stream.timeshift)
```
https://github.com/iptv-org/iptv/blob/62fac8017217747937a2657ff2b2a121f9090954/scripts/commands/playlist/generate.ts#L40

In this week alone, I was made aware of two cases. One where the official stream was bugged and there was no alternative already included in the repo. But more relevantly, a second case where the assumed-canonical stream became bugged, while what is apparently now the official stream was already included in an individual (non-generated) playlist as a third choice listing marked "Not 24/7", and thus, not included in generated playlists.

I think there is a strong case for full generated playlists (at least as a new option). This would help with situations like setup-and-forget where you help non technically capable people who care about just a few channels, and wouldn't mind the duplicates in their favourite list, as long as they are guaranteed non-bugged 24/7 availability for their channels. Okay, maybe not really setup-and-forget, but this will surely require less maintenance.

**Comment by BellezaEmporium** [comment]
This might be why : 

uniqBy is a Lodash function used here to match streams by their name according to their Stream element.
It gets the information from two ways : either the stream channel name, or by the unique ID it is taken from. The unique ID is, by essence, different on each iteration.

The issue might be on the "Stream" element, that would need to be checked for the number of available URLs under the same ID.

As far as I understand it, here's an example : 

- there are 5 "Mochi.us" streams (assuming tvg-id for all streams are as such).
- the playlist system only takes the first iteration from all of them, because "it's high-quality".
- others are ignored because it would bloat the playlists.

Adding a ping system between 2 *insert video file extension* here would allow us to take best answering over best quality, but this won't solve video issues. This cannot be solved by such program in essence.

**Reply by AppleSheeple** [reply]
> Adding a ping system between 2 insert video file extension here would allow us to take best answering over best quality,

The question of *quality* itself is tricky to the point where the HLS standard [added an optional `SCORE` attribute to stream info](
https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-14#section-4.4.6.2) as a *quality* indicator, but didn't bother with trying to properly define *quality*.  The `SCORE` value is described as:

*"an abstract, relative measure of the playback quality-of-experience of the Variant Stream"*

It is that tricky to define *quality* between variants from the same master playlist, despite the number of variables at play being controlled!

Which one is better, a 720p50 stream, or a  1080p25 one?

If two variants have the same resolution and frame-rate, which one is better?

The one with higher bitrate!

Okay. But what if:
  * One variant is i-frames-only or independent-segments, and one is not?
  * One variant is using HEVC and one is using H264?
  * ...


Now imagine the complexity of properly inferring relative *quality* from independent differently sourced streams. With source quality, encoding quality, stability, and availability added as variables!

> but this won't solve video issues. This cannot be solved by such program in essence.

Hence my suggestion to generate unfiltered lists alongside filtered ones. So people who don't mind playlist bloat can make use of alternatives whenever simple assumptions about primary/alternative streams prove to not be holding true!

**Comment by freearhey** [comment]
In my opinion it's better to just remove such laggy broadcasts from the playlist so that everyone can watch the best version by default. As for links labeled as "Not 24/7", sometimes this label is put by mistake and if you remove it, the channel will be ranked again on par with the rest.

**Reply by BellezaEmporium** [reply]
Sadly even the "24/7" stream can be laggy.

For ex: the Asharq News stream is buggy from the base link, but works perfectly from the YouTube alternative, and it's tagged Not 24/7.

Such exceptions including others (that will always occur in every circumstance) will confirm the rule.

**Comment by freearhey** [comment]
Those who want to see all available links in playlists can also use internal playlists directly by clicking the “Raw” button in the upper right corner:

![image](https://github.com/user-attachments/assets/25de97e0-a446-4e13-bcf9-aa78d4eea17c)

**Reply by AppleSheeple** [reply]
OP is specifically talking about ***generated*** playlists, genius.

**Reply by freearhey** [reply]
I'm aware of that.

**Comment by freearhey** [comment]
Now the same result can be achieved simply by adding a unique Feed ID to the link, for example:

```
AsharqNews.sa@YouTube
AsharqNews.sa@HEVC
```

In this case, both streams will go into the generated playlists.

**Reply by AppleSheeple** [reply]
That's cool and useful. But a fully unfiltered list would still be useful, unless all streams are getting unique Feed IDs, which I'm assuming is not the case.

**Reply by BellezaEmporium** [reply]
> That's cool and useful. But a fully unfiltered list would still be useful, unless all streams are getting unique Feed IDs, which I'm assuming is not the case.

Assuming the tvg-id is the same **but** the feed ID is different, this will count as two different streams. Unless you all call them AsharqNews.sa@HEVC in this case, it might try and pick one of them.

**Comment by freearhey** [comment]
Raw versions (without any filtering or sorting) of the playlists are now available at this link:

```sh
https://iptv-org.github.io/iptv/raw/<FILENAME>.m3u
```
